#!/bin/bash

cd ..

# conda installation
./genetic_data/scripts//install_conda.sh

# Tools installation
./genetic_data/scripts//install_HTSeq.sh
./genetic_data/scripts//install_bowtie2.sh
./genetic_data/scripts//install_fastp.sh
./genetic_data/scripts//install_featureCounts.sh
./genetic_data/scripts//install_hisat2.sh
./genetic_data/scripts//install_kallisto.sh
./genetic_data/scripts//install_salmon.sh
./genetic_data/scripts//install_sra_tools.sh
./genetic_data/scripts//install_star.sh
./genetic_data/scripts//set_bilattice_count.sh

# Download all the reads
./genetic_data/reads/seqcB/get_seqcB_data.sh
./genetic_data/reads/seqcB/clean_seqcB_data.sh
./genetic_data/reads/beers_all_bias/get_BEERS_reads.sh
./genetic_data/reads/rat/get_rat_data.sh
./genetic_data/reads/seqcA/get_seqcA_data.sh
./genetic_data/reads/seqcA/clean_seqA_data.sh

# Downlaod all the genomes
./genetic_data/genomes/get_homo_sapiens_genome.sh
./genetic_data/genomes/get_mus_musculus_genome.sh
./genetic_data/genomes/get_rattus_norvegicus_genome.sh

# Download all the transcriptomes
./genetic_data/transcriptomes/get_rattus_norvegicus_transcriptome.sh
./genetic_data/transcriptomes/get_homo_sapiens_transcriptome.sh
./genetic_data/transcriptomes/get_mus_musculus_transcriptome.sh

# Download all the annotations
./genetic_data/annotations/get_homo_sapiens_annotation.sh
./genetic_data/annotations/create_mirna_rattus_norvegicus_tools_annotation.sh
./genetic_data/annotations/get_rattus_norvegicus_annotation.sh
./genetic_data/annotations/get_mus_musculus_annotation.sh
./genetic_data/annotations/create_mirna_rattus_norvegicus_annotation.sh

# Create all the indexes
./genetic_data/indexes/bowtie2_index_homo_sapiens/create_bowtie2_index_homo_sapiens.sh
./genetic_data/indexes/kallisto_index_mus_musculus/create_kallisto_index_mus_musculus.sh
./genetic_data/indexes/salmon_index_mus_musculus/create_salmon_index_mus_musculus.sh
./genetic_data/indexes/bowtie2_index_mus_musculus/create_bowtie2_index_mus_musculus.sh
./genetic_data/indexes/salmon_index_homo_sapiens/create_salmon_index_homo_sapiens.sh
./genetic_data/indexes/kallisto_index_homo_sapiens/create_kallisto_index_homo_sapiens.sh
./genetic_data/indexes/bowtie2_index_rattus_norvegicus/create_bowtie2_index_rattus_norvegicus.sh
./genetic_data/indexes/star_index_mus_musculus/create_star_index_mus_musculus.sh
./genetic_data/indexes/star_index_homo_sapiens/create_star_index_homo_sapiens.sh
./genetic_data/indexes/kallisto_index_rattus_norvegicus/create_kallisto_index-rattus_norvegicus.sh
./genetic_data/indexes/star_index_rattus_norvegicus/create_star_index_rattus_norvegicus.sh

# Create all the alignments
./genetic_data/alignments/bowtie2_seqcA_all/run_bowtie2_seqcA_all.sh
./genetic_data/alignments/bowtie2_seqcB_all/run_bowtie2_seqcB_all.sh
./genetic_data/alignments/star_rat_all/run_star_rat_all.sh
./genetic_data/alignments/star_seqcB_all/run_star_seqcB_all.sh
./genetic_data/alignments/star_beers_all_bias_all/run_star_beers_all_bias_all.sh
./genetic_data/alignments/star_seqcA_all/run_star_seqcA_all.sh
./genetic_data/alignments/bowtie2_beers_all_bias_all/run_bowtie2_beers_all_bias_all.sh
./genetic_data/alignments/bowtie2_rat_all/run_bowtie2_rat_all.sh./genetic_data/alignments/bowtie2_seqcA_all/run_bowtie2_seqcA_all.sh
./genetic_data/alignments/bowtie2_seqcB_all/run_bowtie2_seqcB_all.sh
./genetic_data/alignments/star_rat_all/run_star_rat_all.sh
./genetic_data/alignments/star_seqcB_all/run_star_seqcB_all.sh
./genetic_data/alignments/star_beers_all_bias_all/run_star_beers_all_bias_all.sh
./genetic_data/alignments/star_seqcA_all/run_star_seqcA_all.sh
./genetic_data/alignments/bowtie2_beers_all_bias_all/run_bowtie2_beers_all_bias_all.sh
./genetic_data/alignments/bowtie2_rat_all/run_bowtie2_rat_all.sh

# Compute all the counts
./genetic_data/counts/star_featureCounts_all_seqcA/run_star_featureCounts_all_seqcA.sh
./genetic_data/counts/star_bilatticeCount_rat/run_star_bilatticeCount_rat.sh
./genetic_data/counts/star_bilatticeCount_rat/run_star_bilatticeCount_rat_schweizer.sh
./genetic_data/counts/star_bilatticeCount_rat/run_star_bilatticeCount_rat_hamacher.sh
./genetic_data/counts/salmon_beers_all_bias/run_salmon_beers_all_bias.sh
./genetic_data/counts/bowtie2_featureCounts_all_seqcA/run_bowtie2_featureCounts_all_seqcA.sh
./genetic_data/counts/star_HTSeq_all_seqcA/run_star_HTSeq_all_seqcA.sh
./genetic_data/counts/salmon_seqcB/run_salmon_seqcB.sh
./genetic_data/counts/kallisto_seqcA/run_kallisto_seqcA.sh
./genetic_data/counts/kallisto_seqcB/run_kallisto_seqcB.sh
./genetic_data/counts/bowtie2_HTSeq_all_seqcA/run_bowtie2_HTSeq_all_seqcA.sh
./genetic_data/counts/salmon_seqcA/run_salmon_seqcA.sh
./genetic_data/counts/star_HTSeq_beers_all_bias/run_star_HTSeq_beers_all_bias.sh
./genetic_data/counts/star_HTSeq_rat/run_star_HTSeq_rat.sh
./genetic_data/counts/bowtie2_bilatticeCount_beers_all_bias/run_bowtie2_bilatticeCount_beers_all_bias_schwarzer.sh
./genetic_data/counts/bowtie2_bilatticeCount_beers_all_bias/run_bowtie2_bilatticeCount_beers_all_bias.sh
./genetic_data/counts/bowtie2_bilatticeCount_beers_all_bias/run_bowtie2_bilatticeCount_beers_all_bias_hamacher.sh
./genetic_data/counts/bowtie2_bilatticeCount_all_seqcB/run_bowtie2_bilatticeCount_all_seqcB_hamacher.sh
./genetic_data/counts/bowtie2_bilatticeCount_all_seqcB/run_bowtie2_bilatticeCount_all_seqcB_schwarzer.sh
./genetic_data/counts/bowtie2_bilatticeCount_all_seqcB/run_bowtie2_bilatticeCount_all_seqcB.sh
./genetic_data/counts/kallisto_beers_all_bias/run_kallisto_beers_all_bias.sh
./genetic_data/counts/bowtie2_featureCounts_rat/run_bowtie2_featureCounts_rat.sh
./genetic_data/counts/star_bilatticeCount_all_seqcA/run_star_bilatticeCount_all_seqcA_schwarzer.sh
./genetic_data/counts/star_bilatticeCount_all_seqcA/run_star_bilatticeCount_all_seqcA.sh
./genetic_data/counts/star_bilatticeCount_all_seqcA/run_star_bilatticeCount_all_seqcA_hamacher.sh
./genetic_data/counts/bowtie2_bilatticeCount_all_seqcA/run_bowtie2_bilatticeCount_all_seqcA_hamacher.sh
./genetic_data/counts/bowtie2_bilatticeCount_all_seqcA/run_bowtie2_bilatticeCount_all_seqcA.sh
./genetic_data/counts/bowtie2_bilatticeCount_all_seqcA/run_bowtie2_bilatticeCount_all_seqcA_schwarzer.sh
./genetic_data/counts/bowtie2_HTSeq_beers_all_bias/run_bowtie2_HTSeq_beers_all_bias.sh
./genetic_data/counts/bowtie2_HTSeq_all_seqcB/run_bowtie2_HTSeq_all_seqcB.sh
./genetic_data/counts/star_HTSeq_all_seqcB/run_star_HTSeq_all_seqcB.sh
./genetic_data/counts/bowtie2_HTSeq_rat/run_bowtie2_HTSeq_rat.sh
./genetic_data/counts/kallisto_rat/run_kallisto_rat.sh
./genetic_data/counts/star_bilatticeCount_all_seqcB/run_star_bilatticeCount_all_seqcB.sh
./genetic_data/counts/star_bilatticeCount_all_seqcB/run_star_bilatticeCount_all_seqcB_hamacher.sh
./genetic_data/counts/star_bilatticeCount_all_seqcB/run_star_bilatticeCount_all_seqcB_schwarzer.sh
./genetic_data/counts/bowtie2_bilatticeCount_rat/run_bowtie2_bilatticeCount_rat.sh
./genetic_data/counts/bowtie2_bilatticeCount_rat/run_bowtie2_bilatticeCount_rat_schweizer.sh
./genetic_data/counts/bowtie2_bilatticeCount_rat/run_bowtie2_bilatticeCount_rat_hamacher.sh
./genetic_data/counts/star_featureCounts_all_seqcB/run_star_featureCounts_all_seqcB.sh
./genetic_data/counts/star_bilatticeCount_beers_all_bias/run_star_bilatticeCount_beers_all_bias.sh
./genetic_data/counts/star_bilatticeCount_beers_all_bias/run_star_bilatticeCount_beers_all_bias_hamacher.sh
./genetic_data/counts/star_bilatticeCount_beers_all_bias/run_star_bilatticeCount_beers_all_bias_schwarzer.sh
./genetic_data/counts/star_featureCounts_beers_all_bias/run_star_featureCounts_beers_all_bias.sh
./genetic_data/counts/bowtie2_featureCounts_all_seqcB/run_bowtie2_featureCounts_all_seqcB.sh
./genetic_data/counts/bowtie2_featureCounts_beers_all_bias/run_bowtie2_featureCounts_beers_all_bias.sh
./genetic_data/counts/star_featureCounts_rat/run_star_featureCounts_rat.sh

# Transform the outputs
./genetic_data/outputs/get_rt_pcr_data.sh
./genetic_data/outputs/get_beers_ground_truth.sh
conda activate bilattice_env
python genetic_data/outputs/transform_counts.py
